### DISCUSSION TITLE
Plan for retirement of rgdal, rgeos

### DISCUSSION THREAD

**Discussion by Jean-Romain** [r-lidar/lidR#471] [Ideas]
## Tagging interested persons

 i.e. developers of packages that depend on lidR

@ptompalski @bi0m3trics @tgoodbody @carlos-alberto-silva @Blecigne @oliviermartin7 @DRAAlmeida @Molina-Valero @jmmonnet 

[aRchi](https://github.com/umr-amap/aRchi) [FORTLS](https://github.com/Molina-Valero/FORTLS) [leafR](https://github.com/DRAAlmeida/leafR) [lidaRtRee](https://gitlab.irstea.fr/jean-matthieu.monnet/lidaRtRee) [weblidar-treetop](https://github.com/carlos-alberto-silva/weblidar-treetop) [viewshed3d](https://github.com/Blecigne/viewshed3d) [rGEDI](https://github.com/carlos-alberto-silva/rGEDI) [spanner](https://github.com/bi0m3trics/spanner) [lidRmetrics](https://github.com/ptompalski/lidRmetrics)

## Context

`rgdal` and `rgeos` will be retired on Jan 1st 2024.

- https://twitter.com/RogerBivand/status/1407705212538822656
- https://www.youtube.com/watch?v=cK08bxUJn5A

`raster` and `sp` are based on `rgdal`/`rgeos`. `lidR` is based on `raster` and `sp` because it was created before `sf`, `terra` and `stars`. This means that sooner or later `lidR` will run into trouble (actually it is more or less already the case). So we must fully move to `sf`, `terra`/`stars` without depending on `sp` and `raster`. Do do so we need to:

- design new functions that return `sf`/`terra` object in replacement of  function that return `sp`/`raster` objects.
- redesign the `LAS*` classes such as they no longer inherit `sp::Spatial`. This point being by far the most complex task and it will **necessarily break backward compatibility**

In this thread, I'm putting down my plan to fully remove dependency to `sp`/`raster` in 2024.  The plan is to do it with minimum backward incompatibilities and a lot of time to update the workflow.

**Warning**: I'm putting here draft of ideas because it helps me to make a plan. It is in no way a definitive plan  and anybody reading this is welcome to contribute to this thread.

## Replacement functions

Former functions that return `sp`/`raster` will be supported until it causes issues on CRAN i.e. maybe Jan 1st 2024 maybe later

| Function        | Replacement      |
| ------------- |:-------------:|
| `grid_metrics`     |  `pixel_metrics` |
| `grid_terrain`     | `rasterize_terrain`      |
| `grid_canopy` |  `rasterize_canopy`      |
| `grid_density` |  `rasterize_density`      |
| `tree_metrics` |  `crown_metrics`     |
| `delineate_tree` |    `crown_metrics`   |
|`find_trees` | `locate_trees`  |
| `track_sensor`|   `track_sensor`  |

## LAS class

The current `LAS` definition is:

```r
setClass(
  Class = "LAS", contains = "Spatial",
  representation(data = "data.table", header = "LASheader", index = "list")
)
```
Which implies it has `@proj4string` and `@bbox` slots. To minimize backward incompatibilities we need something like:

```r
setClass(
  Class = "LAS",
  representation(data = "data.table", header = "LASheader", index = "list", crs = "crs")
)
```

This way all user code that uses `las@data$Z` instead of `las$Z` will still work. What will fail however is `las@proj4string` or `las@bbox` which is unlikely to be used. Notice that `bbox` is not useful because it is redundant with the header, it was there only by inheritance. Also function `projection` and `crs` will still return `sp::CRS` object even if they are outdated in the R spatial ecosystem for backward compatibility.

## LAScatalog class

The `LAScatalog` definition is

```r
setClass(
  Class = "LAScatalog",
  contains = "SpatialPolygonsDataFrame",
  representation(
    chunk_options = "list",
    processing_options = "list",
    output_options = "list",
    input_options = "list",
    index = "list"
  )
)
```

The best option would be to become an `sf` object instead of an `S4` classes but this is inconsistent with the `LAS` class. One well backward compatible option is

```r
setClass(
  Class = "LAScatalog",
  representation(
    data = "sf"
    chunk_options = "list",
    processing_options = "list",
    output_options = "list",
    input_options = "list",
    index = "list"
  )
)
```

This will still provide access to e.g. `ctg@data[["Min.Z"]]` or `ctg$GUID` without modifying the code. Other operation being unlikely to be performed by user. Only by internal code

## LAScluster class

The `LAScluster` class definition is

```r
setClass(
  Class = "LAScluster", contains = "Spatial",
  representation(
    center = "list",
    bbbox  = "matrix",
    width  = "numeric",
    height = "numeric",
    buffer = "numeric",
    shape  = "numeric",
    select = "character",
    filter = "character",
    wkt    = "character",
    files  = "character",
    name   = "character",
    save   = "character",
    alt_dir = "character",
    index = "list"
  )
)
```

But being an internal class it is easier to deal with it

## New functions and functions that need an update

- New: `st_crs` ; `st_crs<-` for `LAS` and `LAScatalog`
- Change: `projection` ; `crs` ;  `projection<-` ; `crs<-`
- Remove: `spTransform`

**Comment by DRAAlmeida** [comment]
Dear Jean, thanks a lot for your email.

Cheers,

**Comment by Jean-Romain** [comment]
Update.

I created a version of `lidR v4` that no longer uses `sp` at all. The code is 100% `sf` based using `sf::as_Spatial()` to return outputs preserving backward compatibility. Inheritances to spatial objects has been removed with only minor backward incompatibilities (few call that are not expected to be used). Tables comparing before/after compatibility are reported below. All CRAN packages depending on `lidR` were tested without problem. `sp` is now suggested only.

I will now try to remove `raster` which will be another game but that should not affect users. The major problem was the inheritance to `sp` objects.

----------------------

Backward compatibility using:

```r
library(lidR)
```

Note: some calls that do not work are likely to work loading e.g. `sf`. For example `st_crs()` did not work in v3 but worked with `library(sf)` because it becomes equivalent to `sf::st_crs()`.

- **Call**: is the called function
- **LAS**: applied to a `LAS` object
- **LAScatalog**: applied to a `LAScatalog` object
- **LASheader**: applied to a `LASheader` object
- **LAScluster**: applied to a `LAScluster` object
-  **Regression**: is there a regression
- :x:/:heavy_check_mark: compares v3/v4

## CRS

| Call | LAS | LAScatalog | LASheader | LAScluster | Regression |
|:-----|:---:|:----------:|:---------:|:----------:|:----------:|
| `st_crs` | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark: | |
| `sf::st_crs` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark: | |
| `st_crs<-` | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :x:/:x: | |
| `sf::st_crs<-` | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :x:/:x: | |
| `projection` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark: | |
| `raster::projection` | :heavy_check_mark:/:x:  | :heavy_check_mark:/:x:  | :x:/:x:  | :heavy_check_mark:/:x: | :triangular_flag_on_post: |
| `projection<-` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark: | |
| `raster::projection<-` | :heavy_check_mark:/:x:  | :heavy_check_mark:/:x:  | :x:/:x:  | :heavy_check_mark:/:x: | :triangular_flag_on_post: |
| `crs` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark: | |
| `raster::crs` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark: | |
| `sf::st_is_longlat` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark: | |

## Area

| Call | LAS | LAScatalog | LASheader | LAScluster | Regression |
|:-----|:---:|:----------:|:---------:|:----------:|:----------:|
| `st_bbox` | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark: | |
| `sf::st_bbox` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:heavy_check_mark:  |  :heavy_check_mark:/:heavy_check_mark: | |
| `extent` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:| |
| `raster::extent` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark: | |
| `bbox` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark: | |
| `sp::bbox` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark: | |
| `st_area` | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :x:/:x: | |
| `sf::st_area` | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :x:/:x: | |
| `area` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:x: | |
| `raster::area` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:x: | |

## Accessor

| Call | LAS | LAScatalog | LASheader | LAScluster | Regression |
|:-----|:---:|:----------:|:---------:|:----------:|:----------:|
| `[` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:x: | :x:/:x: | |
| `$` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:heavy_check_mark:  | :x:/:x: | |
| `[[` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark: | :x:/:heavy_check_mark:  | :x:/:x: | |
| `@data$` | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/:heavy_check_mark:  | :x:/:x: | :x:/:x: | |
| `@proj4string` | :heavy_check_mark:/:x:  | :heavy_check_mark:/:x:  | :x:/:x: | :heavy_check_mark:/:x: | :triangular_flag_on_post: |

## R base

| Call | LAS | LAScatalog | LASheader | LAScluster | Regression |
|:-----|:---:|:----------:|:---------:|:----------:|:----------:|
| `nrow`/`ncol` | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark: | :x:/:x: | :x:/:x: | |
| `dim` | :x:/:heavy_check_mark:  | :x:/:heavy_check_mark: | :x:/:x: | :x:/:x: | |
| `rbind`  | :heavy_check_mark:/:heavy_check_mark:  | :heavy_check_mark:/ :heavy_check_mark: | :x:/:x: | :x:/:x: | |
| `names`  | :x: :heavy_check_mark:  | :x:/:heavy_check_mark: | :x:/:x: | :x:/:x: | |

## Others

| Call | LAS | LAScatalog | LASheader | LAScluster | Regression |
|:-----|:---:|:----------:|:---------:|:----------:|:----------:|
| `st_convex_hull` | :x:/:heavy_check_mark:  | :x:/:x:  | :x:/:x:  | :x:/:x: | |
| `sf::st_convex_hull` | :x:/:heavy_check_mark:  | :x:/:x:  | :x:/:x:  | :x:/:x: | |
